### PR TITLE
Export: avoid pointless image<->Python data conversions

### LINKED ISSUE #790: High res textures as height for bump node cause massive freeze
**Description**
When having a material with a bump node, which takes a high resolution texture as input (in my tests they were 8k textures) and exporting as gltf 2.0, the (or at least my) PC slowly hangs up and almost crashes. I thought it might be my CPU. which couldn't keep up with the exporter, so i tried to let it sit for some time, with no results (at one point i have waited 90 minutes), but it might also be that the RAM isn't sufficient (when looking at blender in the task manager, its RAM use raises from ~500mb to almost 5gb, where it stops and freezes the whole computer, and even closes other programs).
This problem can be prevented by disabling "Material" under "Geometry" in the export setting, but it would still be great if this problem didnt exist in the first place.

**How to reproduce**
1. Create any type of 3D mesh object, and add a material
2. Create a bump node, and connect it to the normal input of the principled node
3. Create a new texture in the image panel with a high resolution (i tested with 8192x)
4. create an image texture node, select the newly created image, and connect it with the 'height' slot of the bump node
5. export the scene as gltf (with materials)
**CAUTION: This might lead your PC to freeze for half an hour or longer!**

**Expected behavior**
The Exporter creates the primitives of the mesh with the material (see console output) and then freezes. The RAM usage of blender raises tremendously, until the whole computer freezes.

**Version**
 - OS: Windows 10, Version 1903
 - Blender Version 2.80 / 2.81

My CPU is an Intel Ceon E3-1231 v3, and i have 8GB RAM, if those stats matter.

### PR DESCRIPTION
**Pull request by scurest**
Fixes #790.

Take two on #798 m(_ _)m

Change ExportImage to store a description of how its image should be built instead of storing an array of pixel data. This lets use make more intelligent decisions when we encode, namely only doing Image<->Python data conversions when necessary for assembling an image out of channels. This gives a nice win in export time and memory use (~twice as fast for models with big textures).

Here's a list of how this works vs master

|Case|master|This PR|
|:---:|:---:|:---:|
|Image needs to be assembled out of channels|Images converted to Python data, channels packed, converted back to an Image, saved and read back|Same as master|
|Existing Image that isn't on disk (non-dirty)|Image converted to Python data, not changed, converted back to an Image, Image saved and read back|Image copied with `image.copy`, saved and read back|
|Existing Image that isn't on disk (dirty)|Same as if non-dirty|Same as master, except it avoids going through the numpy array part of the conversion|
|Existing Image that is on disk|Image converted to Python data, file on disk read back|File on disk read back|

There should be no changes from master in any of the exported images except what's described in https://github.com/KhronosGroup/glTF-Blender-IO/pull/798#issuecomment-561284590.

The only real difference with #798 is in __encode_from_image. I abandoned the attempt to save a copy of an image, and just copied the whole image and saved that normally. This should be much easier to check correctness of (except for one thing.....)

### CODE REVIEW

File: addons/io_scene_gltf2/blender/exp/gltf2_blender_image.py [hunk 1]
```diff
@@ -14,121 +14,198 @@
...
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        tmpfilename = tmpdirname + '/img'
+        tmp_image.filepath_raw = tmpfilename
 
-            with open(tmpfilename, "rb") as f:
-                encoded_image = f.read()
+        # NOT A TYPO!!! If you delete this line, the
+        # assignment on the next line will not work.
+        tmp_image.file_format
+        tmp_image.file_format = file_format
```
Comment (scurest) [inline addons/io_scene_gltf2/blender/exp/gltf2_blender_image.py]: Believe it or not.... :)

insanity

I can reproduce with both today's 2.82 alpha and 2.79.

Other than that, this is the same way master saves an image.

**Comment by scurest** [comment]
Has anyone looked at this?

**Comment by emackey** [comment]
@julienduroure Hopefully this can make it upstream in time for 2.82.  The stable branch is [scheduled to be created Jan 9th](https://devtalk.blender.org/t/30-december-2019/11068).  This is worthy of the "critical fix" tag in my humble opinion.

**Comment by julienduroure** [comment]
@emackey yep, don't worry, I will be back home before 9th,  so it will be merged on time

**Pull request merged**